### PR TITLE
Feature/syal-180

### PR DESCRIPTION
**Pull request by MaksimR-AVISPL**

### CODE REVIEW

**Review by ShervinM-AVISPL — CHANGES_REQUESTED** [review]

File: src/main/java/com/avispl/dal/communicator/cisco/CiscoCommunicator.java [hunk 1]
```diff
@@ -0,0 +1,2554 @@
...
+    private ExtendedStatistics localStatistics;
+    private EndpointStatistics localEndpointStatistics;
+
+    Unmarshaller xmlUnmarshaller;
+
+    /**
+     * Instantiate {@link CiscoCommunicator}
+     */
+    public CiscoCommunicator() {
+        setTrustAllCertificates(true);
```
[the remark this answers is not in the document]
Comment (MaksimR-AVISPL) [reply src/main/java/com/avispl/dal/communicator/cisco/CiscoCommunicator.java:325]: Yeah, for most of them. Pretty sure it's not the case for XiO or other communicators that depend on a remote cloud api, but for all the devices with http/https protocols - yes :)

File: src/main/java/com/avispl/dal/communicator/cisco/CiscoCommunicator.java [hunk 2]
```diff
@@ -0,0 +1,2554 @@
...
+     * @param statistics map to set statistics data to
+     * @param controls list of controls to add controllable properties to
+     * @param status device status data fetched from {@link #statusPath}
+     * @param configuration device configuration data fetched from {@link #configurationPath}
+     */
+    private void populateRoomAnalyticsData(Map<String, String> statistics, List<AdvancedControllableProperty> controls, CiscoConfiguration configuration, CiscoStatus status){
+        RoomAnalytics roomAnalytics = status.getRoomAnalytics();
+        if (roomAnalytics != null) {
+            addStatisticsParameter(statistics, "RoomAnalytics#PeoplePresence", roomAnalytics.getPeoplePresence());
+            ValueSpaceRefHolder peoplePresenceDetectorState = configuration.getRoomAnalytics().getPeoplePresenceDetector();
```
[the remark this answers is not in the document]
Comment (MaksimR-AVISPL) [reply src/main/java/com/avispl/dal/communicator/cisco/CiscoCommunicator.java:1422]: well, basically it is expected that it's there if it's present in status, but just to be on the safe side - yes, it's better to have the check

File: src/main/java/com/avispl/dal/communicator/cisco/CiscoCommunicator.java [hunk 3]
```diff
@@ -0,0 +1,2749 @@
...
+                contentChannelStats.setPacketLossTx(Integer.valueOf(netstat.getLoss()));
+                break;
+            default:
+                logger.info("Channel direction not supported: " + channel.getDirection());
+                break;
+        }
+    }
+
+    private Integer extractValueInt(String value) {
+        if(isNumeric(value)) {
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/CiscoCommunicator.java:2074]: Missing '!'

File: src/main/java/com/avispl/dal/communicator/cisco/CiscoCommunicator.java [hunk 4]
```diff
@@ -0,0 +1,2749 @@
...
+     * @param parameterName name of the parameter to set
+     * @param value value of the parameter to fetch value and min/max thresholds from as well as the current param value
+     */
+    private void addStatisticsParameterWithSlider(Map<String, String> statistics, List<AdvancedControllableProperty> controllableProperties,
+                                                  String parameterName, ValueSpaceRefHolder value) {
+        if(value != null) {
+            String parameterValue = value.getValue();
+            List<Object> minMaxValues = value.getValues();
+            if (isNumeric(parameterValue) && !minMaxValues.isEmpty()
+                    && (value.getType().equals(ValueSpaceRefHolderType.INT) || value.getType().equals(ValueSpaceRefHolderType.VS_INT))) {
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/CiscoCommunicator.java:2600]: value.getType() is being extracted multiple times.

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/AudioConfigurationCommandType.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright (c) 2021 AVI-SPL Inc. All Rights Reserved.
+ */
+package com.avispl.dal.communicator.cisco.controller.type;
+
+public enum AudioConfigurationCommandType {
```
[the remark this answers is not in the document]
Comment (ShervinM-AVISPL) [reply src/main/java/com/avispl/dal/communicator/cisco/controller/type/AudioConfigurationCommandType.java:15]: Enum members need brief javadoc, or there will be warnings. For example, MicrophoneMode could have simple java doc that looks like /** Represents the mode of microphone */

Here's an example of an enum that does not show any warnings:
 ```suggestion
public enum SimulationMode {
		/** few devices in a call, few devices violating thresholds. here 10% of devices should be in a call and another 10% violating thresholds */
		regular,
		/** few devices in a call, large number of devices violating thresholds. here 10% of devices should be in a call and another 60% violating thresholds */
		high_threshold_violation,
		/**
		 * few devices in a call, few devices violating thresholds, connection to cloud is simulated as down. here 10% of devices should be in a call and
		 * another 10% violating thresholds
		 */
		cloud_down
	}

```

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/AudioControlCommandType.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+ * Audio control command type class.
+ * Serves the purpose of distinguishing between the audio control command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum AudioControlCommandType {
+    Volume
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/AudioControlCommandType.java:16]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/CameraConfigurationCommandType.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+ * Serves the purpose of distinguishing between the camera config command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum CameraConfigurationCommandType {
+    CameraFlip, CameraIrSensor, CameraMirror, CameraWhitebalanceLevel, CameraWhitebalanceMode, CameraGammaLevel, CameraGammaMode,
+    CameraFocusMode, CameraBrightnessDefaultLevel, CameraBrightnessMode, CameraBacklightDefaultMode, TriggerAutofocusPreset
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/CameraConfigurationCommandType.java:17]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/CameraControlCommandType.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ * Camera control command type class.
+ * Serves the purpose of distinguishing between the camera control command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum CameraControlCommandType {
+    CameraPan, CameraTilt, CameraZoom, CameraFocus
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/CameraControlCommandType.java:16]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/ConferenceConfigurationCommandType.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+ * Conference configuration command type class.
+ * Serves the purpose of distinguishing between the conference config command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum ConferenceConfigurationCommandType {AutoAnswerDelay, AutoAnswerMode, AutoAnswerMute, DefaultCallProtocol,
+    DefaultCallRate, DoNotDisturbTimeout, EncryptionMode, FarEndControl}
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/ConferenceConfigurationCommandType.java:16]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/NetworkServicesConfigurationCommandType.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+ * Network services configuration command type class.
+ * Serves the purpose of distinguishing between the network services config command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum NetworkServicesConfigurationCommandType {
+    H323Mode, SIPMode, TelnetMode, SSHMode, CDPMode, UPnPMode, NTPMode, WelcomeText
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/NetworkServicesConfigurationCommandType.java:16]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/ProximityConfigurationCommandType.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+ * Proximity configuration command type class.
+ * Serves the purpose of distinguishing between the camera proximity command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum ProximityConfigurationCommandType {
+    ProximityMode, CallControl, ContentShareToClients, ContentShareFromClients
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/ProximityConfigurationCommandType.java:16]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/RoomAnalyticsConfigurationCommandType.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+ * Room analytics configuration command type class.
+ * Serves the purpose of distinguishing between the room analytics config command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum RoomAnalyticsConfigurationCommandType {
+    PeoplePresenseMode
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/RoomAnalyticsConfigurationCommandType.java:16]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/StandbyConfigurationCommandType.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+ * Standby configuration command type class.
+ * Serves the purpose of distinguishing between the standby config command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum StandbyConfigurationCommandType {
+    BootAction, Control, Delay, StandbyAction, WakeupAction
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/StandbyConfigurationCommandType.java:16]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/UserInterfaceConfigurationCommandType.java [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+ * Serves the purpose of distinguishing between the user interface config command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum UserInterfaceConfigurationCommandType {
+    WelcomeText, Language, KeyTones, ContactInfoType, FeaturesCallEnd, FeaturesMidCallControls, FeaturesCallStart, ShareStart,
+    WhiteboardStart, HideAllFeatures, AwakeBrandingColors, IncomingCallNotification, OSDOutput
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/UserInterfaceConfigurationCommandType.java:17]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/VideoConfigurationCommandType.java [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+ * Serves the purpose of distinguishing between the video config command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum VideoConfigurationCommandType {
+    DefaultMainSource, DefaultPresentationSource, Monitors, Type, Visibility, PresentationSelection, Quality, CECMode, MonitorRole,
+    OverscanLevel, Resolution
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/VideoConfigurationCommandType.java:17]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/controller/type/VideoSelfViewConfigurationCommandType.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+ * Video selfview configuration command type class.
+ * Serves the purpose of distinguishing between the video selfview config command types, to have the right
+ * instance of the payload built for a specific control command.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum VideoSelfViewConfigurationCommandType {
+    DefaultFullscreenMode, DefaultMode, DefaultOnMonitorRole, DefaultPIPPosition, OnCallDuration, OnCallMode
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/controller/type/VideoSelfViewConfigurationCommandType.java:16]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/dto/ValueSpaceRefHolderType.java [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+
+/**
+ * Possible values for a valuespace node types
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
+public enum ValueSpaceRefHolderType {
+    VS_INT, INT, STR, TTPAR, DEFAULT
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/dto/ValueSpaceRefHolderType.java:14]: Enum members javadoc missing

File: src/main/java/com/avispl/dal/communicator/cisco/dto/ValueSpaceRefHolder.java [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+ * or
+ * <Connector item="2" maxOccurrence="2"></Connector>
+ *
+ * valueSpaceRef is later used for fetching additional value details from /valuespace.xml, to define control limits,
+ * controllable property type etc.
+ *
+ * @author Maksym.Rossiitsev / Symphony Dev Team<br>
+ * Created on Apr 26, 2021
+ * @since 1.0
+ */
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/dto/ValueSpaceRefHolder.java:24]: special characters such as '<' will result in Symphony's javadoc generator to fail. You should use '&lt;'.
Or, you can just wrap any special characters in {@code ..... }, for example:
{@code <DviType valueSpaceRef="/Valuespace/TTPAR_DviType">AutoDetect</DviType>}

File: src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/call/CallDisconnectCommand.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * Retrieves {@code {@link #disconnectCommand}}
+     *
+     * @return value of {@link #disconnectCommand}
+     */
+    public Disconnect getDisconnectCommand() {
+        return disconnectCommand;
+    }
+
+    public CallDisconnectCommand() {
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/call/CallDisconnectCommand.java:49]: public method missing javadoc

File: src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/call/response/CallDisconnectResult.java [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+public class CallDisconnectResult {
+    @XmlAttribute(name = "status")
+    private String status;
+
+    public String getStatus() {
+        return status;
+    }
+
+    public void setStatus(String status) {
+        this.status = status;
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/call/response/CallDisconnectResult.java:27]: public methods missing javadocs

File: src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/standby/StandbyCommand.java [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+    @XmlElement(name = "Deactivate")
+    private Deactivate deactivate;
+
+    public StandbyCommand(Activate activate) {
+        this.activate = activate;
+    }
+
+    public StandbyCommand(Deactivate deactivate) {
+        this.deactivate = deactivate;
+    }
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/standby/StandbyCommand.java:72]: public constructors missing javadocs

File: src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/systemunit/SystemUnitCommand.java [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    public static class Boot {
+        @XmlAttribute
+        private String command = "True";
+        @XmlElement(name = "Action")
+        private String action;
+
+        public Boot(){}
+        public Boot(BootAction action) {
+            this.action = action.name();
+        }
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/systemunit/SystemUnitCommand.java:48]: public constructors missing javadocs

File: src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/userinterface/UserInterfaceCommand.java [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+                @XmlElement(name = "Text")
+                private ValueSpaceRefHolder text;
+                @XmlElement(name = "Duration")
+                private ValueSpaceRefHolder duration;
+
+                public Display() {}
+                public Display(String text, Integer duration) {
+                    this.text = new ValueSpaceRefHolder(text);
+                    this.duration = new ValueSpaceRefHolder(duration);
+                }
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/userinterface/UserInterfaceCommand.java:61]: public constructors missing javadocs

File: src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/userinterface/UserInterfaceCommand.java [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+            }
+        }
+        @XmlElement(name = "TextLine")
+        private TextLine textLine;
+
+        public Message () {}
+        public Message (String text, Integer duration) {
+            textLine = new TextLine();
+            textLine.setDisplay(new TextLine.Display(text, duration));
+        }
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/dto/control/commands/userinterface/UserInterfaceCommand.java:136]: public constructors missing javadocs.
Please check all constructors have javadocs (sorry for such tedious tasks, but if we don't do it, we'll have too many warnings)

File: src/main/java/com/avispl/dal/communicator/cisco/dto/valuespace/ValueSpace.java [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+package com.avispl.dal.communicator.cisco.dto.valuespace;
+
+import javax.xml.bind.annotation.*;
+
+/***
+ * This class represents final values that contain specific attributes, needed for the proper data processing.
+ * This is used for processing configuration parameters with multiple values potentially available. Such
+ * data is dynamic so we can't rely on static values set in the code, nor we can rely on anything manually
+ * configured. So, every time a specific node it processed, like:
+ *      <Protocol valueSpaceRef="/Valuespace/TTPAR_CallProtocol">Auto</Protocol>
```
Comment (ShervinM-AVISPL) [inline src/main/java/com/avispl/dal/communicator/cisco/dto/valuespace/ValueSpace.java:13]: special characters will fail on javadoc generation. Wrap then in {@code ..... }

**Review by ShervinM-AVISPL — CHANGES_REQUESTED** [review]

**Review by ShervinM-AVISPL — APPROVED** [review]
Looks good @MaksimR-AVISPL. Please don't forget to format CiscoCommunicator class before you merge, I saw a lot of little formatting issues.

**Pull request merged**